contracts/marketpools.js

```diff
@@ -0,0 +1,346 @@
+    return false;
+  }
+  return true;
+}
+
+async function validatePool(tokenPair) {
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  const revPool = await api.db.findOne('pools', { tokenPair: [quoteSymbol, baseSymbol].join(':') });
+  if (!api.assert(pool === null && revPool === null, 'a pool already exists for this tokenPair')) return false;
+  return true;
+}
+
+async function updatePoolStats(pool, baseAdjusted, quoteAdjusted, swap = true) {
+  const uPool = pool;
+  // precise quantities are needed here for K calculation
+  uPool.baseQuantity = api.BigNumber(pool.baseQuantity).plus(baseAdjusted);
+  uPool.quoteQuantity = api.BigNumber(pool.quoteQuantity).plus(quoteAdjusted);
+  // remainder are statistical and can be rounded (updated for swaps only)
+  uPool.basePrice = api.BigNumber(uPool.quoteQuantity).dividedBy(uPool.baseQuantity).toFixed(pool.precision);
+  uPool.quotePrice = api.BigNumber(uPool.baseQuantity).dividedBy(uPool.quoteQuantity).toFixed(pool.precision);
+  if (swap) {
+    uPool.baseVolume = api.BigNumber(uPool.baseVolume).plus(Math.abs(baseAdjusted)).toFixed(pool.precision);
+    uPool.quoteVolume = api.BigNumber(uPool.quoteVolume).plus(Math.abs(quoteAdjusted)).toFixed(pool.precision);
+  }
+  await api.db.update('pools', uPool);
+}
+
+actions.createPool = async (payload) => {
+  const {
+    tokenPair, isSignedWithActiveKey,
+  } = payload;
+
+  // get contract params
+  const params = await api.db.findOne('params', {});
+  const { poolCreationFee } = params;
+
+  // eslint-disable-next-line no-template-curly-in-string
+  const utilityTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'" });
+
+  const authorizedCreation = api.BigNumber(poolCreationFee).lte(0) || api.sender === api.owner
+    ? true
+    : utilityTokenBalance && api.BigNumber(utilityTokenBalance.balance).gte(poolCreationFee);
+
+  if (api.assert(authorizedCreation, 'you must have enough tokens to cover the creation fee')
+    && await validateTokenPair(tokenPair)
+    && await validatePool(tokenPair)
+    && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')) {
+    const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+    const baseToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: baseSymbol });
+    const quoteToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: quoteSymbol });
+    const newPool = {
+      tokenPair,
+      baseQuantity: 0,
+      baseVolume: 0,
+      basePrice: 0, // Base per Quote (usual way of reading a pair)
+      quoteQuantity: 0,
+      quoteVolume: 0,
+      quotePrice: 0, // Quote per Base (reverse price)
+      precision: Math.min(baseToken.precision, quoteToken.precision),
```

Comment:
Why min here vs max? Or perhaps default should be the highest available? I'm not sure, and will defer to you. But since this appears to be used in intermediate computations throughout the code, larger is better.

Reply:
Original idea was use min for all intermediate computations and when sending out at the end use that for both tokens. But since that we are using individual token precision now on the final send so I believe this could be max now. In the process of reviewing the parts with rounding.